Adding slab metrics

ipstatic opened the pull request:
This resolves #2 by porting @jjneely PR from the original exporter ([PR here](https://github.com/snapbug/memcache_exporter/pull/7)). Please note I did have to rename two metrics as they would collide with the item names being added (items_evicted_total and items_reclaimed_total). If this is an issue please let me know.

juliusv commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
 			nil,
 			nil,
 		),
+		active_slabs: prometheus.NewDesc(
```
`active_slabs` -> `activeSlabs`

juliusv commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+		),
+		itemsEvicted: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "items_evicted_total"),
+			"Total of times an item had to be evicted from the LRU before it expired.",
```
Total of times -> Total number of times

juliusv commented on README.md:
```diff
@@ -40,16 +42,52 @@ The exporter collects a number of statistics from the server:
-# TYPE memcached_items_evicted_total counter
-# HELP memcached_items_reclaimed_total Number of times an entry was stored using memory from an expired entry.
-# TYPE memcached_items_reclaimed_total counter
+# HELP memcached_evicted_total Number of valid items removed from cache to free memory for new items.
```
Ah ok, the new ones have an additional `slab` label, so not exactly the same semantics...

juliusv commented on main.go:
```diff
@@ -204,6 +357,52 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+			}
+			ch <- prometheus.MustNewConstMetric(e.slabsCommands, prometheus.CounterValue, parse(v, "cas_badval"), slab, "cas", "badval")
+
+			slab_set := math.NaN()
```
`slab_set` -> `slabSet`

juliusv commented:
:+1: besides comments, but it would be a bonus if someone had a short look at it who knows something about memcached.

grobie commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+			nil,
+		),
+		itemsNumber: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "items_number_total"),
```
Instead of renaming these metrics, I'd use a common "subsystem" for all the metrics coming from the slab+items statistics commands, probably called `slab` or the plural `slabs`. So all metrics with a slab label would have the common prefix `memcached_slab_*`. Memcached uses the term `number` for the slab identifier, so I'd use that as label name instead of `slab`.

This applies to all slab related metrics.

grobie commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
 			nil,
 		),
+		active_slabs: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "active_slabs_total"),
```
This is a gauge and should not have the `_total` suffix. Also the current number of slabs can be calculated with `count by(number)(memcached_slab_xxx)`, I would not export this as extra metric.

grobie commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+			nil,
+		),
+		itemsEvicted: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "items_evicted_total"),
```
With my suggestion above, this would become `memcached_slab_items_evicted_total{number="x"}` and not longer conflict with the overall `memcached_items_evicted_total`.

grobie commented on main.go:
```diff
@@ -123,17 +142,131 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+		),
+		itemsNumber: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "items_number_total"),
+			"Total number of items currently stored in this slab class.",
```
This seems to be a gauge and should not mention total in the help text or have the `_total` suffix. Counters must only be used for monotonically increasing values, the current number of items can go up and down.

grobie commented on main.go:
```diff
@@ -204,6 +357,52 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+		for slab, u := range t.Items {
+			slab := strconv.Itoa(slab)
+			ch <- prometheus.MustNewConstMetric(e.itemsNumber, prometheus.CounterValue, parse(u, "number"), slab)
+			ch <- prometheus.MustNewConstMetric(e.itemsAge, prometheus.CounterValue, parse(u, "age"), slab)
```
This is a gauge, as the age can go up and down.

grobie commented on main.go:
```diff
@@ -204,6 +357,52 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+			slab := strconv.Itoa(slab)
+			ch <- prometheus.MustNewConstMetric(e.itemsNumber, prometheus.CounterValue, parse(u, "number"), slab)
+			ch <- prometheus.MustNewConstMetric(e.itemsAge, prometheus.CounterValue, parse(u, "age"), slab)
+			ch <- prometheus.MustNewConstMetric(e.itemsEvicted, prometheus.GaugeValue, parse(u, "evicted"), slab)
```
Check the types of all of these metrics, they seem to be all counters and not gauges (e.g. they'll never go down). A counter should always have the `_total` suffix, a gauge should never have the `_total` suffix.

ipstatic commented:
@juliusv @grobie Thank you very much for your feedback. I believe I have fixed all of your concerns. @grobie the only question in my mind if I did select the correct metric type for each metric. I tried reading the [Memcached protocol doc](https://github.com/memcached/memcached/blob/master/doc/protocol.txt) to grok the type but couldn't tell 100%. Your input would be most appreciated.

grobie commented on main.go:
```diff
@@ -124,16 +142,124 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+			nil,
+		),
+		itemsEvicted: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "slab", "items_evicted"),
```
I'm very sure this and all following `memcached_slab_items_*` metrics are counters and should also have the `_total` suffix. I'll double check our current statistics again, but the wording in the protocol documentation https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L791 "Number of times ..." indicates that as well.

grobie commented on main.go:
```diff
@@ -124,16 +142,124 @@ func NewExporter(server string, timeout time.Duration) *Exporter {
+			nil,
+		),
+		itemsNumber: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "slab", "items_number"),
```
Let's use `memcached_slab_current_items` to be consistent with the global metric name and avoid confusion with the slab number.

grobie commented on main_test.go:
```diff
@@ -64,6 +64,10 @@ func TestAcceptance(t *testing.T) {
 	if err := client.CompareAndSwap(last); err != nil {
 		t.Fatal(err)
 	}
+	large := &memcache.Item{Key: "large", Value: []byte("Hello World this text should be 128 bytes in size so that we may test the slab functionality of memcached. Filler text here to  ")}
```
Cool. Is it possible to add a test for metrics in `slab="2"` as well then?

ipstatic replied:
in my instance the slab number is 5. I don't know if this is due to the size of the text I added or what. Another pair of eyes or someone with more experience with memcache would be appreciated.

ipstatic commented:
@grobie I believe I have fixed all type issues now. Also added more item metrics. I left out the following as I never saw them in my local or production instances of memcache:

- direct_reclaims
- lrutail_reflocked
- moves_within_lru
- number_*
- moves_to_*

jjneely commented:
Thank you @ipstatic !

ipstatic commented:
Also, let me know if I should squash these commits down to a single commit.

grobie commented on README.md:
```diff
@@ -40,16 +40,58 @@ The exporter collects a number of statistics from the server:
+# TYPE memcached_slab_chunks_free_end gauge
+# HELP memcached_slab_chunks_per_page Number of chunks within a single page for this slab class.
+# TYPE memcached_slab_chunks_per_page gauge
+# HELP memcached_slab_chunks_total Total number of chunks allocated to this slab class.
```
I'm still unsure whether this is a counter. While the original `total_chunks` naming could mean that, I have the feeling this is actually a gauge representing the current amount of chunks used. A common naming scheme in Prometheus would be `memcached_slab_current_chunks`. Though, given the ambiguity of the original metric name, let me check some more of our servers.

The latest memcached versions support automatic slab rescaling, I'm trying to find out whether that would cause such numbers to go down again.

ipstatic replied:
If I can help with checking things out, let me know.

grobie commented on README.md:
```diff
@@ -40,16 +40,58 @@ The exporter collects a number of statistics from the server:
+# TYPE memcached_slab_items_tailrepairs_total counter
+# HELP memcached_slab_mem_requested_bytes Number of bytes of memory actual items take up within a slab.
+# TYPE memcached_slab_mem_requested_bytes gauge
+# HELP memcached_slab_pages_total Total number of pages allocated to this slab class.
```
Same as memcached_slab_chunks_total, my gut feeling is this is a gauge and not a counter.

ipstatic commented on main_test.go:
```diff
@@ -80,12 +84,20 @@ func TestAcceptance(t *testing.T) {
+		`memcached_slab_current_items{slab="5"} 1`,
+		`memcached_slab_commands_total{command="set",slab="1",status="hit"} 2`,
+		`memcached_slab_commands_total{command="cas",slab="1",status="hit"} 1`,
+		`memcached_slab_commands_total{command="set",slab="5",status="hit"} 0`,
```
I would think so too but when running the test suite it passes. I haven't really dug into why there isn't a hit. Maybe because the test doesn't try to retrieve the key?

grobie replied:
It'll work when the issue mentioned above is fixed.

grobie commented on main.go:
```diff
@@ -204,6 +374,54 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+
+			slabSet := math.NaN()
+			if slabSetCmd, err := strconv.ParseFloat(v["cmd_set"], 64); err == nil {
+				if cas, casErr := sum(s, "cas_misses", "cas_hits", "cas_badval"); casErr == nil {
```
This needs to be `v`. `s` references the global stats.

ipstatic replied:
Wow, I totally missed this, thanks.

ipstatic commented:
@grobie have you had a change to dig into those two metric names (memcached_slab_chunks_total and memcached_slab_pages_total) and if I should either remove the _total name or change them to a counter instead of a gauge?

grobie commented:
I briefly played around with memcached but I haven't been able to to reduce either of the two values yet. My strong feeling is that these are gauges. As far as I can see, both metrics are of type counter atm, let's go with gauge and remove the _total suffixes. Similar to other metrics exported here, we could name them `memcached_slab_current_(pages|chunks)`, but I'm fine with either.

ipstatic commented:
I went ahead with current_(pages|chunks). Anything else that stands out?

grobie commented:
Merged 94522920b0989620f2e10ba50affc5327306b6e9 into master.

ipstatic commented:
Thank you for all your feedback and patience!
